### DISCUSSION TITLE
Tensor mechanics hardening documents

### DISCUSSION THREAD

**Discussion by dineshp1** [idaholab/moose#21022] [Q&A Modules: General] [answered]
Hi
I am interested in the different hardening schemes implemented in the module. The module is not documented yet so could you please provide some reference documents for hardening schemes?

Thank you,

Regards,
Dinesh P

**Comment by GiudGiud** [comment, marked as answer]
Hello

The module is actually documented. Please see these pages for the base documentation:
https://mooseframework.inl.gov/modules/tensor_mechanics/index.html

And a few hardening related pages:
https://mooseframework.inl.gov/source/materials/IsotropicPlasticityStressUpdate.html
https://mooseframework.inl.gov/source/materials/IsotropicPowerLawHardeningStressUpdate.html

What kind of hardening are you looking for?

Guillaume

**Reply by dineshp1** [reply]
Thank you for the reply. I want to apply linear softening like this:
![image](https://user-images.githubusercontent.com/104110739/168719534-c7d5fa42-883d-4a74-a359-7ffec7f22d46.png)

I found some hardening class implementation(TensorMechanicsHardeningCutExponential) and I can create linear softening now. Although I have a question, what is the internal parameter here? is it strain or something else? 

```
TensorMechanicsHardeningCutExponential::TensorMechanicsHardeningCutExponential(
    const InputParameters & parameters)
  : TensorMechanicsHardeningModel(parameters),
    _val_0(getParam<Real>("value_0")),
    _val_res(parameters.isParamValid("value_residual") ? getParam<Real>("value_residual") : _val_0),
    _intnl_0(getParam<Real>("internal_0")),
    _rate(getParam<Real>("rate"))
{
}

Real
TensorMechanicsHardeningCutExponential::value(Real intnl) const
{
  Real x = intnl - _intnl_0;
  if (x <= 0)
    return _val_0;
  else
    return _val_res + (_val_0 - _val_res) * std::exp(-_rate * x);
}

```

**Reply by WilkAndy** [reply]
The internal parameter depends on your plastic model.  In your case it will be "Plastic shear strain".  Your plastic model (eg, J2, Mohr-Coulomb, etc) must define what "Plastic shear strain" actually is.  If you search through the TensorMechanics test cases you'll find many examples of softening and hardening.  Depending on your model, you may find the piecewise linear function that you are proposing leads to convergence problems due to the non-differentiability (unless you use an explicit time-stepping).  I have attached some documentation - not sure whether it's really what you want

[MOOSE Framework - Hardening and Softening.pdf](https://github.com/idaholab/moose/files/8704855/MOOSE.Framework.-.Hardening.and.Softening.pdf)


[MOOSE Framework - Plasticity and the Return-Mapping Algorithm.pdf](https://github.com/idaholab/moose/files/8704861/MOOSE.Framework.-.Plasticity.and.the.Return-Mapping.Algorithm.pdf)


[MOOSE Framework - Plasticity and the Consistent Tangent Operator.pdf](https://github.com/idaholab/moose/files/8704862/MOOSE.Framework.-.Plasticity.and.the.Consistent.Tangent.Operator.pdf)

**Reply by dineshp1** [reply]
@WilkAndy Thank you so much for the reference docs and for explaining issues with the sharp corner in softening function.
